synapseclient/api/entity_factory.py

```diff
@@ -0,0 +1,391 @@
+        version:            The version number of the entity to retrieve.
+        if_collision: Determines how to handle file collisions. May be:
+
+                - `overwrite.local`
+                - `keep.local`
+                - `keep.both`
+        limit_search:       Limit the search to a specific project or folder. Only used
+            if `synapse_id_or_path` is a path.
+        md5: The MD5 of the file to retrieve. If not passed in, the MD5 will be
+            calculated. Only used if `synapse_id_or_path` is a path.
+        download_file: Whether associated files should be downloaded.
+        download_location: The directory to download the file to.
+        follow_link: Whether to follow a link to its target. This will only do a single
+            hop to the target of the link.
+        entity_to_update: An existing entity class instance to update with data from
+            Synapse.
+        synapse_client: If not passed in or None this will use the last client from
+            the `.login()` method.
+
+        Example: Using this function
+            Download file into cache
+
+                from synapseclient import Synapse
+                from synapseclient.api import get_from_entity_factory
+
+                syn = Synapse()
+                syn.login()
+
+                entity = await get_from_entity_factory(synapse_id_or_path='syn1906479')
+                print(entity.name)
+                print(entity.path)
+
+            Download file into current working directory
+
+                from synapseclient import Synapse
+                from synapseclient.api import get_from_entity_factory
+
+                syn = Synapse()
+                syn.login()
+
+                entity = await get_from_entity_factory(synapse_id_or_path='syn1906479', download_location='.')
+                print(entity.name)
+                print(entity.path)
+
+    Raises:
+        SynapseFileNotFoundError: If the id is not a synapse ID and it is not a valid
+            file path.
+    """
+
+    # If entity is a local file determine the corresponding synapse entity
+    if isinstance(synapse_id_or_path, str) and os.path.isfile(synapse_id_or_path):
+        bundle = await _search_for_file_by_md5(
+            filepath=synapse_id_or_path,
+            limit_search=limit_search,
+            md5=md5,
+            synapse_client=synapse_client,
+        )
+        download_file = False
+
+    elif isinstance(synapse_id_or_path, str) and not utils.is_synapse_id_str(
+        obj=synapse_id_or_path
+    ):
+        raise SynapseFileNotFoundError(
+            (
+                f"The parameter {synapse_id_or_path} is neither a local file path "
+                " or a valid entity id"
+            )
+        )
+    else:
+        synid_and_version = utils.get_synid_and_version(obj=synapse_id_or_path)
+        version = version if version is not None else synid_and_version[1]
+        if version:
```

Comment:
Why do we need this `if version` conditional here? It looks like `version` gets assigned something in the previous line no matter what.

Reply:
It does get assigned, but it can be assigned to `None`. The return type of `get_synid_and_version` is:
`typing.Tuple[str, typing.Union[int, None]]`

Reply:
Thanks for clarifying!